enterprise/server/remote_execution/containers/firecracker/firecracker.go

```diff
@@ -2257,30 +2250,61 @@ func (c *FirecrackerContainer) Exec(ctx context.Context, cmd *repb.Command, stdi
 		c.workspaceStore.EmitUsageMetrics("init")
 	}
 
-	result = c.SendExecRequestToGuest(ctx, cmd, workDir, stdio)
+	// TODO(bduffany): Reuse connection from Unpause(), if applicable
+	conn, err := c.dialVMExecServer(ctx)
+	if err != nil {
+		return commandutil.ErrorResult(status.InternalErrorf("Firecracker exec failed: failed to dial VM exec port: %s", err))
+	}
+	defer conn.Close()
+
+	result = c.SendExecRequestToGuest(ctx, conn, cmd, workDir, stdio)
 	close(execDone)
 
 	ctx, cancel = background.ExtendContextForFinalization(ctx, finalizationTimeout)
 	defer cancel()
 
 	// If FUSE is enabled then outputs are already in the workspace.
 	if c.fsLayout == nil && !*disableWorkspaceSync {
-		// Command was successful, let's unpack the files back to our
-		// workspace directory now.
+		// Unmount the workspace and pause the VM before syncing to ensure that
+		// the guest's FS cache is flushed to the backing file on the host and
+		// that no concurrent operations can occur while we're reading the
+		// workspace disk.
+		//
+		// TODO(bduffany): after notifying users of these warnings below,
+		// make it a hard failure if we fail to unmount or if the workspace disk
+		// is still busy after unmounting.
+		client := vmxpb.NewExecClient(conn)
+		var unmounted bool
+		if rsp, err := client.UnmountWorkspace(ctx, &vmxpb.UnmountWorkspaceRequest{}); err != nil {
+			log.CtxWarningf(ctx, "Failed to unmount workspace")
+		} else {
+			unmounted = true
+			if rsp.GetBusy() {
+				// Do not recycle the VM if the workspace device is still busy after
+				// unmounting.
+				result.DoNotRecycle = true
+				log.CtxWarningf(ctx, "Workspace device is still busy after unmounting - not recycling VM")
+			}
+		}
 		if err := c.pauseVM(ctx); err != nil {
 			result.Error = status.InternalErrorf("error pausing VM: %s", err)
 			return result
 		}
-
 		if err := c.copyOutputsToWorkspace(ctx); err != nil {
 			result.Error = status.WrapError(err, "failed to copy action outputs from VM workspace")
 			return result
 		}
-
+		// Resume the VM and remount the workspace now that we're done copying.
 		if err := c.machine.ResumeVM(ctx); err != nil {
 			result.Error = status.InternalErrorf("error resuming VM after copying workspace outputs: %s", err)
 			return result
 		}
+		if unmounted {
```

Comment:
curious if you're worried about this double-mounting or what? feels to me like there are cases where the error in the unmount() call above doesn't _necessarily_ mean that the guest didn't unmount the drive (even though farfetched).

i guess i'm asking if it would be more or less safe in your opinion to drop this `if unmounted` check.

Reply:
(as an alternative, we could make MountWorkspace a no-op if the workspace is already mounted)

Reply:
hm I think I just forgot to set `DoNotRecycle = true` if the unmount failed (done). That way we don't have to worry about double-mounting because the VM won't be reused.